### PR TITLE
Adds ability to filter nodes if discovery is on

### LINKED ISSUE #303: Specify a type of node to autodiscover
The current method of discovery is pretty broad:
* connect to a node
* call `/_nodes/_all` endpoint (because no nodes were specified)
* add all nodes with an `http_address` value to the servers list. 

The problem is, this requires you to set `http.enabled: false` on the nodes you don't want serving query requests. ( i.e. master nodes). It then follows that you can't get stats from the master nodes, because `http.enabled` is set to false on them, and the http module has to be enabled to get stats out of a node. 

What is needed in this library is a way to specify which nodes you actually want to discover. Unless I missed it, there does seem to be a way to do this if you're constructing an action to get NodesInfo, but the [NodeChecker](https://github.com/searchbox-io/Jest/blob/es_1.0/jest-common/src/main/java/io/searchbox/client/config/discovery/NodeChecker.java#L31) class builds the action completely inside of itself, and there's not a way to specify which nodes to discover from the `HttpClientConfig.Builder`class

What I'm looking for is just a way to specify nodes, so that the endpoint being called will be something like 
```
_nodes/<some node attribute to filter on>/http
``` 
whereas it's currently
```
_nodes/_all/http
```

### PR DESCRIPTION
**Pull request by bdharrington7**
This closes #303 

This added functionality will allow building an HttpClientConfig with an optional filter criteria on nodes info (Previously turning discovery on would seek out all nodes with http endpoints on)

Using the syntax outlined in the manual for [Nodes Info](https://www.elastic.co/guide/en/elasticsearch/reference/current/cluster-nodes-info.html), you can specify which type of nodes to autodiscover based on attributes of each node. 

Note that filtering by `master:true` etc. doesn't work for the Elasticsearch Nodes Info API, but you can use any of the same syntax as in the example, or specify your own custom attributes to filter by, adding values to `node.<some attribute name>: <some attribute value>` in the elasticsearch configuration

### CODE REVIEW

**Comment by bdharrington7** [comment]
I've confirmed that the cornercase I've described is a bug in elasticsearch, which has been resolved in master and slated for (at least) 5.x: https://github.com/elastic/elasticsearch/issues/18100

Waiting to see if they will backport to 2.x

**Comment by bdharrington7** [comment]
There won't be a backport, but I explained how to get desired behavior above when using client nodes

**Comment by bdharrington7** [comment]
@kramer I'm not sure how you backport changes from master to the respective 1.0 / 2.0 branch, but I have a change staged for the 1.0 branch if needed, it only applies the changed files I made in this PR, with some modifications in the tests to accommodate the different ES dependency version

**Comment by bdharrington7** [comment]
Hi @kramer would it be possible to get some comments on this so I can clean it up if needed?

File: .gitignore [hunk 1]
```diff
@@ -1,7 +1,6 @@
 # Common IDEs
 *.iml
 .idea
-.idea/**/*
```
Comment (kramer) [inline .gitignore]: but why?

File: jest/src/main/java/io/searchbox/client/JestClientFactory.java [hunk 2]
```diff
@@ -67,6 +68,9 @@ public JestClient getObject() {
         // set discovery (should be set after setting the httpClient on jestClient)
         if (httpClientConfig.isDiscoveryEnabled()) {
             log.info("Node Discovery enabled...");
+            if (StringUtils.isNotEmpty(httpClientConfig.getDiscoveryFilter())) {
+                log.info("Node Discovery filtering on \"" + httpClientConfig.getDiscoveryFilter() + "\"");
```
Comment (kramer) [inline jest/src/main/java/io/searchbox/client/JestClientFactory.java]: better use the object arguments style i.e.: `log.info("... {0}", arg)`

**Comment by kramer** [comment]
Hi @bdharrington7 

Excuse me for the late reply again...

Changes look good to me apart from the two things I commented above. Could you please add some info about the filtering to the [appropriate section in readme](https://github.com/searchbox-io/Jest/tree/master/jest#node-discovery-through-nodes-api)?

File: .gitignore (hunk 1 above)
Comment (bdharrington7) [inline .gitignore]: Ignoring this file folder is taken care of on line 3; technically it should be .idea/

Comment (kramer) [inline .gitignore]: ah well I missed that ok then :)

**Comment by kramer** [comment]
Thank you for your contribution @bdharrington7 !

**Comment by bdharrington7** [comment]
You're welcome!

**Comment by bdharrington7** [comment]
Hey @kramer Thanks again for merging my PR, Do you have an idea when your next release is? (for the 1.x versions at least?)

**Comment by kramer** [comment]
Probably on 11th or 12th of June.

**Pull request merged**